### PR TITLE
Scopes: Adds kinds for browsing the scope node tree. 

### PR DESCRIPTION
**Pull request by bergquist**
This is my proposal for the Scope Node data model. 

The `ScopeNode` is a tree that can render selected or non selectable items (decided by the `isLeaf` bool). 

Having `leafType` and `leafName` allows the API to include both scopes, dashboards, apps or any other kind within the tree. While we will focus only on scopes right now we should use this for more things eventually. 

`leftName` could have a better name that is more clear about referencing another entity identifier. 

To find the root nodes in the tree I'd like to find all ScopeNode.ParentName that equals "" or nil depending on if the name can be a pointer or not. I think a pointer is a cleaner solution but Im not sure if it would be possible to filter based on it. 

As for the API server part... I've just tried to copy as much as possible from the code I see for `scope` and `scopedashboard` 😆

### CODE REVIEW

**Review by toddtreece — APPROVED** [review]

File: pkg/apis/scope/v0alpha1/types.go [hunk 1]
```diff
@@ -73,3 +73,27 @@ type ScopeDashboardBindingList struct {
...
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type ScopeNode struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec ScopeNodeSpec `json:"spec,omitempty"`
+}
+
+type ScopeNodeSpec struct {
+	ParentName string `json:"parentName"`
```
[the remark this answers is not in the document]
Comment (bergquist) [reply pkg/apis/scope/v0alpha1/types.go]: What if I want to query for ScopeNodes with `parentName == nil` ?

[the remark this answers is not in the document]
Comment (toddtreece) [reply pkg/apis/scope/v0alpha1/types.go]: > Sorry for my lack of knowledge about the apiserver

no need to apologize for not knowing something! i'm not sure of the correct approach here either 😄 

maybe using an empty string would work?
```go
func SelectableScopeNodeFields(obj *scope.ScopeNode) fields.Set {
        parentName := ""
        if obj.Spec.ParentName != nil {
          parentName = *obj.Spec.ParentName
        }
	return generic.MergeFieldsSets(generic.ObjectMetaFieldsSet(&obj.ObjectMeta, false), fields.Set{
		"spec.parentName": parentName,
	})
}
```

[the remark this answers is not in the document]
Comment (bergquist) [reply pkg/apis/scope/v0alpha1/types.go]: Yupp! that worked :)

File: pkg/apis/scope/v0alpha1/types.go [hunk 2]
```diff
@@ -73,3 +73,30 @@ type ScopeDashboardBindingList struct {
...
+type ScopeNode struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec ScopeNodeSpec `json:"spec,omitempty"`
+}
+
+type ScopeNodeSpec struct {
+	//+optional
+	ParentName   *string `json:"parentName"`
```
Comment (kylebrandt) [inline pkg/apis/scope/v0alpha1/types.go:87]: Does this need to be a pointer? Would empty string struffice for "No parent", or does null and empty have different meanings?

**Pull request merged**